### PR TITLE
MRG: Use sphinx-gallery, add API page

### PR DESCRIPTION
**Pull request by Eric89GXL**
This PR:

1. Switches to sphinx-gallery
2. Adds API page for class/function doc (I've always wanted this)
3. Cleans up some docstrings
4. Adds intersphinx linking
5. Fixes some OSX DYLD_LIBRARY_PATH silliness (proper fix for #171)

These were necessary to get things looking good on OSX.

### CODE REVIEW

**Comment by agramfort** [comment]
@Eric89GXL can you share a rendered site to see?

**Comment by Eric89GXL** [comment]
@agramfort finally here:

https://ci.appveyor.com/api/buildjobs/o8v5r82fmmgbjhw0/artifacts/doc%2F_build%2Fhtml.zip

**Comment by agramfort** [comment]
looks great

**Comment by mwaskom** [comment]
Thanks for taking the initiative here, Eric. The new "modern" css looks mostly good. Here are some thoughts:

- The box on the index page with the logo/name/tagline (I think web devs call this a "hero" element?) feels too large for its relatively uninteresting visual content.
- The tagline "Python library for visualizing cortical surface representations of neuroimaging data" feels a little dull, I guess because it starts with "Python library" which is obvious from the name, and is a little wordy. Elsewhere we have used "Cortical neuroimaging visualization in Python" which is a little snappier and emphasizes the specific application of the library a little better. But it can be improved.
- The bootswatch theme in general is clean and nice, but the lack of any color on pages other than the homepage feels very bland.
- By moving the first sentence of the intro description into the tagline, we start with "PySurfer is primarily intended for use with Freesurfer, " which does not grab my attention! This paragraph needs some improvement.
- I don't like how we've lost the links to other parts of the site from the main text on the homepage.
- I just noticed that our install page still says "does not work on Python 3", which is no longer True. Also do we have a hard dependency on IPython still?
- "Documentation" in the header should probably be "Tutorial" or something more specific, since everything is really "documentation". Although the stuff in there isn't a very good tutorial.... Also "Examples" should maybe be "Gallery".
- Images in the example gallery are not working (at least in the static Appveyor download).

**Comment by Eric89GXL** [comment]
> The box on the index page with the logo/name/tagline (I think web devs call this a "hero" element?) feels too large for its relatively uninteresting visual content.

> The bootswatch theme in general is clean and nice, but the lack of any color on pages other than the homepage feels very bland.

> By moving the first sentence of the intro description into the tagline, we start with "PySurfer is primarily intended for use with Freesurfer, " which does not grab my attention! This paragraph needs some improvement.

Jumbotron :)

With these aesthetic choices we risk getting into a back-and-forth bikeshed-type argument because there is no clear right answer. So I'll take a stab at addressing these comments, but I'd like to stop at whatever is the minimal acceptable documentation improvement, and iterate over subsequent PRs (probably not by me) to tweak such aesthetic / layout / wordings further. Okay?

> I don't like how we've lost the links to other parts of the site from the main text on the homepage.

I disagree on this point. They are redundant with the links at the top. If we put them on the main page, they should really have identical text / naming as the top links. So unless the links at the top need some additional explanation that you could put to the right of such main-body links (and I don't think they do), I don't see a reason to also have them on the main page.

> I just noticed that our install page still says "does not work on Python 3", which is no longer True. Also do we have a hard dependency on IPython still?

Technically Python 3 is possible if you use non-standard Anaconda repos. I won't add that for now (can be a separate PR because the wording might be tricky), but I'll remove the bit about IPython.

> "Documentation" in the header should probably be "Tutorial" or something more specific, since everything is really "documentation". Although the stuff in there isn't a very good tutorial.... Also "Examples" should maybe be "Gallery".

I'll change these.

> Images in the example gallery are not working (at least in the static Appveyor download).

Yes AppVeyor only does the "noplot" build.

**Comment by mwaskom** [comment]
>  So I'll take a stab at addressing these comments, but I'd like to stop at whatever is the minimal acceptable documentation improvement, and iterate over subsequent PRs (probably not by me) to tweak such aesthetic / layout / wordings further. Okay?

Well ... sure ... but I don't see the point of merging a PR that is aesthetically/textually a regression from the current docs. Unless I'm missing something, adding automatically generated API docs and moving the example gallery to sphinx-gallery is orthogonal to using sphinx-bootswatch and altering the homepage text. So I think it's fine for a PR about the former not to get bogged down in a discussion of the latter, but maybe they should be separate PRs/discussions.

**Comment by mwaskom** [comment]
> I disagree on this point. They are redundant with the links at the top. If we put them on the main page, they should really have identical text / naming as the top links. So unless the links at the top need some additional explanation that you could put to the right of such main-body links (and I don't think they do), I don't see a reason to also have them on the main page.

They are not redundant because they serve different purposes. The links at the top are navigational elements for people who know what they are looking for. The main text on the homepage should tell people what there *is* to look for. I agree that the latter goal is better served by a bit of narration about the different elements of the website than just a list of links. You can look at the seaborn homepage to get an idea of what my preference would be.

**Comment by Eric89GXL** [comment]
> I don't see the point of merging a PR that is aesthetically/textually a regression from the current docs.

I agree, but I didn't realize you considered these (and I don't consider these) aesthetic changes as regressions. If you think of them the changes way, they should be separated so I'll revert.

**Comment by mwaskom** [comment]
If we end up sticking with the sphinx_bootstrap/bootswatch themes, one thing I hate about the CSS is the giant text in the API parameter descriptions; in the seaborn docs I [make it smaller](https://github.com/mwaskom/seaborn/blob/master/doc/_static/style.css#L3).

**Comment by mwaskom** [comment]
> I agree, but I didn't realize you considered these (and I don't consider these) aesthetic changes as regressions. If you think of them the changes way, they should be separated so I'll revert.

Perhaps its just my familiarity (and authorship) bias, but the new site with the bland color palette and reduce text feels very unfinished and unwelcoming. I think it can be made better than the original docs with some judicious changes, even simple things like including a small version of the log in the header. I recognize that bikeshedding about aesthetics can be annoying, but you are kind of inviting it when you propose a whole-sale restyling of the docs...

**Comment by Eric89GXL** [comment]
> I recognize that bikeshedding about aesthetics can be annoying, but you are kind of inviting it when you propose a whole-sale restyling of the docs...

Yeah this is true. I wrongly / naively assumed that folks would agree with me main thrust of the changes. I have now reverted them in any case so we can deal with them separately, so this should now hopefully just be functional stuff switching to sphinx-gallery etc. now.

**Comment by Eric89GXL** [comment]
New build:

https://ci.appveyor.com/api/buildjobs/g7a95ykjrgaqunaj/artifacts/doc%2F_build%2Fhtml.zip

Travis is just waiting for OSX to build, but should come back green.

**Comment by Eric89GXL** [comment]
CIs are happy, ready for review/merge from my end.

FWIW talking to @agramfort he didn't like the modern look either!

File: doc/Makefile [hunk 1]
```diff
@@ -3,130 +3,115 @@
...
+	@echo "Build finished. The HTML pages are in _build/html_stable."
 
-html:
-	../bin/pysurfer --help > documentation/pysurfer_usage.txt
-	$(SPHINXBUILD) -b html $(ALLSPHINXOPTS) $(BUILDDIR)/html
+html_dev:
+	BUILD_DEV_HTML=1 $(SPHINXBUILD) -D raise_gallery=1 -D abort_on_example_error=1 -b html $(ALLSPHINXOPTS) _build/html
 	@echo
-	@echo "Build finished. The HTML pages are in $(BUILDDIR)/html."
+	@echo "Build finished. The HTML pages are in _build/html"
```
Comment (agramfort) [inline doc/Makefile]: why changing this? this is standard sphinx template to use $(BUILDDIR)

File: doc/conf.py [hunk 2]
```diff
@@ -91,26 +117,44 @@
...
+
 
 
-# -- Options for HTML output ---------------------------------------------------
+# -- Options for HTML output ----------------------------------------------
 
 # The theme to use for HTML and HTML Help pages.  See the documentation for
 # a list of builtin themes.
-html_theme = "sphinxdoc"
+# html_theme = 'bootstrap'
```
Comment (agramfort) [inline doc/conf.py]: remove

File: doc/conf.py [hunk 3]
```diff
@@ -91,26 +117,44 @@
...
+#     'navbar_pagenav': False,
+#     'navbar_class': "navbar",
+#     'bootstrap_version': "3",  # default
+#     'navbar_links': [
+#         ("Install", "install"),
+#         ("Documentation", "documentation/index"),
+#         ("API", "python_reference"),
+#         ("Examples", "auto_examples/index"),
+#     ],
+#     }
```
Comment (agramfort) [inline doc/conf.py]: remove

File: doc/documentation/custom_viz.rst [hunk 4]
```diff
@@ -25,15 +25,15 @@ should either be a single number to make a square window, or a pair of
 values, ``(width, height)``, to make a rectangular window.
 
 Changing the curvature color scheme
------------------------------------ 
+-----------------------------------
 
-By default, a new :class:`Brain` instance displays the binarized 
+By default, a new :class:`surfer.Brain` instance displays the binarized
```
Comment (agramfort) [inline doc/documentation/custom_viz.rst]: why do you need this ?

you could say to use the surfer namespace

**Comment by mwaskom** [comment]
I tried building the docs (`make html_stable`) but had some problems. A few examples complained about not finding the fsaverage BA labels -- I think this is because Freesurfer 6.0 added "exvivo" to their names. The second issue is crashing out of quite a few examples because it's looking for the SUBJECTS_DIR in `/Applications/freesurfer/subjects`. But that's not where my SUBJECTS_DIR is, which the $SUBJECTS_DIR variable in my terminal confirms. Is there some reason that the changes make pysurfer think it's running on a mac? `git grep Applications` is empty...

**Comment by mwaskom** [comment]
Here are the exceptions:

```

Exception occurred:
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/sphinx_gallery/gen_gallery.py", line 280, in sumarize_failing_examples
    "\n" + "-" * 79)
ValueError: Here is a summary of the problems encountered when running the examples

Unexpected failing examples:
/home/mwaskom/code/pysurfer/examples/plot_resting_correlations.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/sphinx_gallery/gen_rst.py", line 472, in execute_code_block
    exec(code_block, example_globals)
  File "<string>", line 8, in <module>
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/surfer/viz.py", line 416, in __init__
    subjects_dir = _get_subjects_dir(subjects_dir=subjects_dir)
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/surfer/utils.py", line 659, in _get_subjects_dir
    % subjects_dir)
ValueError: The subjects directory /Applications/freesurfer/subjects does not exist.


/home/mwaskom/code/pysurfer/examples/plot_transparent_brain.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/sphinx_gallery/gen_rst.py", line 472, in execute_code_block
    exec(code_block, example_globals)
  File "<string>", line 16, in <module>
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/surfer/viz.py", line 416, in __init__
    subjects_dir = _get_subjects_dir(subjects_dir=subjects_dir)
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/surfer/utils.py", line 659, in _get_subjects_dir
    % subjects_dir)
ValueError: The subjects directory /Applications/freesurfer/subjects does not exist.


/home/mwaskom/code/pysurfer/examples/plot_label.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/sphinx_gallery/gen_rst.py", line 472, in execute_code_block
    exec(code_block, example_globals)
  File "<string>", line 13, in <module>
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/surfer/viz.py", line 1289, in add_label
    % filepath)
ValueError: Label file /usr/local/freesurfer/current/subjects/fsaverage/label/lh.BA1.label does not exist


/home/mwaskom/code/pysurfer/examples/plot_topographic_contours.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/sphinx_gallery/gen_rst.py", line 472, in execute_code_block
    exec(code_block, example_globals)
  File "<string>", line 13, in <module>
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/surfer/viz.py", line 416, in __init__
    subjects_dir = _get_subjects_dir(subjects_dir=subjects_dir)
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/surfer/utils.py", line 659, in _get_subjects_dir
    % subjects_dir)
ValueError: The subjects directory /Applications/freesurfer/subjects does not exist.


/home/mwaskom/code/pysurfer/examples/plot_probabilistic_label.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/sphinx_gallery/gen_rst.py", line 472, in execute_code_block
    exec(code_block, example_globals)
  File "<string>", line 10, in <module>
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/surfer/viz.py", line 416, in __init__
    subjects_dir = _get_subjects_dir(subjects_dir=subjects_dir)
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/surfer/utils.py", line 659, in _get_subjects_dir
    % subjects_dir)
ValueError: The subjects directory /Applications/freesurfer/subjects does not exist.


-------------------------------------------------------------------------------
The full traceback has been saved in /tmp/sphinx-err-oK1cAm.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
Makefile:42: recipe for target 'html_stable' failed
make: *** [html_stable] Error 1
```

File: examples/plot_meg_inverse_solution.py [hunk 5]
```diff
@@ -13,18 +13,16 @@
...
 print(__doc__)
 
-path = os.path.dirname(__file__)
-
 """
 define subject, surface and hemisphere(s) to plot
 """
 subject_id, surf = 'fsaverage', 'inflated'
-hemi = 'split'
+hemi = 'lh'  # 'split'
```
Comment (agramfort) [inline examples/plot_meg_inverse_solution.py]: why these changes in this file?

File: doc/Makefile (hunk 1 above)
Comment (Eric89GXL) [inline doc/Makefile]: I copied this from MNE, need to fix in both places I guess

File: doc/conf.py (hunk 3 above)
Comment (Eric89GXL) [inline doc/conf.py]: I wanted to leave these to make it easy for others to try

**Comment by agramfort** [comment]
can you revert the doc/_static/banner.png changes?

File: examples/plot_meg_inverse_solution.py (hunk 5 above)
Comment (Eric89GXL) [inline examples/plot_meg_inverse_solution.py]: It was broken when I tried to build (the multi-view backend is flakey)

**Comment by Eric89GXL** [comment]
I can revert the banner but it seemed useful to leave it transparent ?

File: examples/plot_meg_inverse_solution.py (hunk 5 above)
Comment (agramfort) [inline examples/plot_meg_inverse_solution.py]: can you clean up then?

**Comment by agramfort** [comment]
initial banner looks nicer. I seemed to see visual artifacts on circles.

File: doc/conf.py (hunk 3 above)
Comment (agramfort) [inline doc/conf.py]: we'll redo it when needed

**Comment by Eric89GXL** [comment]
Ahh I must have done a bad job of making it transparent. I'll revert it.

File: doc/documentation/custom_viz.rst (hunk 4 above)
Comment (Eric89GXL) [inline doc/documentation/custom_viz.rst]: Where?

File: examples/plot_meg_inverse_solution.py (hunk 5 above)
Comment (Eric89GXL) [inline examples/plot_meg_inverse_solution.py]: I left this in intentionally to "suggest" to the user they could do it, but sure

File: doc/documentation/custom_viz.rst (hunk 4 above)
Comment (agramfort) [inline doc/documentation/custom_viz.rst]: at the top of the rst file you can add 

.. currentmodule:: surfer

**Comment by Eric89GXL** [comment]
> Is there some reason that the changes make pysurfer think it's running on a mac? git grep Applications is empty...

Shouldn't be, but I can check. Are you on Linux or Windows?

File: examples/plot_probabilistic_label.py [hunk 6]
```diff
@@ -16,6 +16,7 @@
 
 print(__doc__)
 
+environ['SUBJECTS_DIR'] = '/Applications/freesurfer/subjects'
```
Comment (christianbrodbeck) [inline examples/plot_probabilistic_label.py]: I think this is why it "thinks it's on a mac"

Comment (Eric89GXL) [inline examples/plot_probabilistic_label.py]: Whoops @mwaskom this is the cruft that caused problems

**Comment by Eric89GXL** [comment]
Comments addressed. @mwaskom build problems should be fixed, feel free to try again.

**Comment by mwaskom** [comment]
I still get one error on Freesurfer 6.0

```
Exception occurred:
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/sphinx_gallery/gen_gallery.py", line 280, in sumarize_failing_examples
    "\n" + "-" * 79)
ValueError: Here is a summary of the problems encountered when running the examples

Unexpected failing examples:
/home/mwaskom/code/pysurfer/examples/plot_label.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/sphinx_gallery/gen_rst.py", line 472, in execute_code_block
    exec(code_block, example_globals)
  File "<string>", line 40, in <module>
  File "/home/mwaskom/anaconda2/lib/python2.7/site-packages/surfer/viz.py", line 1289, in add_label
    % filepath)
ValueError: Label file /usr/local/freesurfer/current/subjects/fsaverage/label/lh.entorhinal.label does not exist

```

**Comment by mwaskom** [comment]
In the header there is a link to the "modules" page, which isn't that useful for PySurfer. Could that be changed to link directly to the `python_reference.html` page?

**Comment by mwaskom** [comment]
I bet this is not related to the doc changes, but I just noticed in the gallery that the negative side of Overlays has a backwards colorbar. @christianbrodbeck is this related to your recent PR do you think?

**Comment by christianbrodbeck** [comment]
I did modify some overlay code, I'll check, which example?

**Comment by mwaskom** [comment]
It's apparent in the gallery in the `plot_fmri_volume` example which shows the negative side of the overlay colormap, and you can see it partway through running the `plot_resting_correlations` example.

**Comment by christianbrodbeck** [comment]
Yes I can confirm, I'll work on a fix

**Comment by Eric89GXL** [comment]
> In the header there is a link to the "modules" page, which isn't that useful for PySurfer. Could that be changed to link directly to the python_reference.html page?

Maybe, but I don't have time to look,let's do it in another PR. For now I've reverted the change that added those links.

I also fixed the label / build problem.

**Comment by mwaskom** [comment]
> Maybe, but I don't have time to look,let's do it in another PR. For now I've reverted the change that added those links.

Sounds good — our API is a little weird in that pretty much everything you'd want to know is handled by methods on `Brain` (sigh) so it's worth thinking about how to best organize the API docs so people can get that information with minimal clicking. Totally fine with punting on those details, though.

**Comment by Eric89GXL** [comment]
CIs are happy, let me know if there's anything else needed for this one.

**Comment by mwaskom** [comment]
Looks good for now, thanks!
